src/partition/repart.c

```diff
@@ -3280,12 +3282,70 @@ static int do_make_directories(Partition *p, const char *root) {
         return 0;
 }
 
-static int partition_populate(Partition *p, const char *node) {
+static int partition_populate_directory(Partition *p, char **ret_root, char **ret_tmp_root) {
+        _cleanup_free_ char *root = NULL;
+        int r;
+
+        assert(ret_root);
+        assert(ret_tmp_root);
+
+        /* When generating squashfs, we need the directory root to be available when we call squashfs.
+         * Because we might have multiple source trees, we build a temporary source tree beforehand where we
+         * merge all our inputs. We then use this merged source tree to create the squashfs filesystem. */
+
+        if (!streq(p->format, "squashfs")) {
+                *ret_root = NULL;
+                *ret_tmp_root = NULL;
+                return 0;
+        }
+
+        /* If we only have a single directory that's meant to become the root directory of the filesystem,
+         * we can shortcut this function and just use that directory as the root directory instead. If we
+         * allocate a temporary directory, it's stored in `ret_tmp_root` to indicate it should be removed.
+         * Otherwise, we return the directory to use in `root` to indicate it should not be removed. */
+
+        if (strv_length(p->copy_files) == 2 && strv_length(p->make_directories) == 0 && streq(p->copy_files[1], "/")) {
+                _cleanup_free_ char *s = NULL;
+
+                s = strdup(p->copy_files[0]);
+                if (!s)
+                        return log_oom();
+
+                *ret_root = TAKE_PTR(s);
+                *ret_tmp_root = NULL;
+                return 0;
+        }
+
+        r = mkdtemp_malloc("/var/tmp/repart-XXXXXX", &root);
+        if (r < 0)
+                return log_error_errno(r, "Failed to create temporary directory: %m");
+
+        r = do_copy_files(p, root);
+        if (r < 0)
+                return r;
+
+        r = do_make_directories(p, root);
+        if (r < 0)
+                return r;
+
+        r = fsync_path_at(AT_FDCWD, root);
+        if (r < 0)
+                return log_error_errno(r, "Failed to synchronize temporary directory: %m");
```

Comment:
these are the same three function calls as in the case where we populate ourselves. Maybe split this out into a helper call that we call from both places?

Reply:
actually, fsyncing is kinda pointless here... we don't care if this hits the disk. We just need it to be there long enough for mksquashfs to pick it up. The fsync is counter productive here. please drop